### PR TITLE
include token & currency transfers

### PR DESCRIPTION
**Pull request by ilinzweilin**
- added nftApproval  to issue()
- added currencyApproval to repay()
- added nft tranfer from borrower to proxy on issue()
- added currency transfer from borrower to proxy on repay()
- added nft transfer from proxy to borrower on unlock()

### CODE REVIEW

File: src/actions.sol [hunk 1]
```diff
@@ -23,17 +27,28 @@ contract OperatorLike {
     function relyInvestor(address usr) public;
 }
 
-contract Actions {
+contract Actions is DSTest{
```
Comment (xmxanuel) [inline src/actions.sol]:
```suggestion
contract Actions {
```

File: src/actions.sol [hunk 2]
```diff
@@ -50,30 +64,34 @@ contract Actions {
...
+        erc20.approve(address(shelf), amount);
+        // transfer money from borrower to proxy
+        erc20.transferFrom(msg.sender, address(this), amount);
         shelf.repay(loan, amount);
     }
 
-    function unlock(ShelfLike shelf, uint loan) public {
+    function unlock(ShelfLike shelf, NFTLike registry, uint token, uint loan) public {
         shelf.unlock(loan);
+        registry.transferFrom(address(this), msg.sender, token);
```
Comment (xmxanuel) [inline src/actions.sol:81]: here theoretically we could allow more flexibility by providing an `usr` instead of `msg.sender`

**Review by xmxanuel — APPROVED** [review]
LGTM.

the only downside with the approve approach is:

lets say somebody transfer the ownership of the accesstoken NFT. The
new owner could for example use existing erc20 approvals to take some currency from the previous owner.

File: src/actions.sol [hunk 3]
```diff
@@ -50,30 +64,34 @@ contract Actions {
         borrowWithdraw(shelf, loan, amount, usr);
     }
 
-    function issueLockBorrowWithdraw(ShelfLike shelf, address registry, uint token, uint amount, address usr) public {
-        uint loan = shelf.issue(registry, token);
+    function transferIssueLockBorrowWithdraw(ShelfLike shelf, NFTLike registry, uint token, uint amount, address usr) public {
```
Comment (xmxanuel) [inline src/actions.sol:67]: We could rename the `borrow` in the shelf to something else and rename this entire action to `borrow`.

Comment (ilinzweilin) [reply src/actions.sol:67]: yes, I was thinking about that. I guess too late because of the audit

**Comment by ilinzweilin** [comment]
@Manuel we can make the borrower always approve just the repayment amount. would result in an additional transaction for each repayment, same as transferring money to the proxy on every repayment

File: src/actions.sol (hunk 3 above)
Comment (charleenfei) [reply src/actions.sol:67]: i agree, it's confusing to explain the difference between `withdraw` and `borrow`. but at the very least, maybe we could rename these methods to something a bit shorter? like, maybe just `withdraw`, and then add comments on what is going on?

Comment (ilinzweilin) [reply src/actions.sol:67]: its a combination of all the functions that are executed in the shelf... considering that we have different combinations, I like this pattern. Shorter would be nicer but how would you diffrenciate between issueLockBorrowWithdraw and lockBorrowWithdraw

Comment (charleenfei) [reply src/actions.sol:67]: just also thinking about tinlake.js 😂

Comment (charleenfei) [reply src/actions.sol:67]: `issueWithdraw` v just `withdraw`?

**Review by charleenfei — APPROVED** [review]

File: src/actions.sol (hunk 3 above)
Comment (lucasvo) [reply src/actions.sol:67]: none of this is really user frontend, so the naming is fine I think :)

File: src/actions.sol (hunk 2 above)
Comment (lucasvo) [reply src/actions.sol:81]: we can add that later.

**Review by lucasvo — APPROVED** [review]

**Pull request merged**